File: python/data_structures/hash_map.py
"""
Hash Map (with chaining)
------------------------
Insert:  O(1) average, O(n) worst
Lookup:  O(1) average, O(n) worst
Delete:  O(1) average, O(n) worst

Rolling my own hash map to understand how it works under the hood.
Uses separate chaining (linked lists per bucket) for collision handling.
Resizes when load factor exceeds 0.75.

Obviously in real Python code you'd just use a dict.
"""

INITIAL_CAPACITY = 16
LOAD_FACTOR_THRESHOLD = 0.75


class HashMap:
    def __init__(self, capacity=INITIAL_CAPACITY):
        self._capacity = capacity
        self._size = 0
        self._buckets = [[] for _ in range(capacity)]

    def put(self, key, value):
        idx = self._hash(key)
        bucket = self._buckets[idx]

        # update if key already exists
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket[i] = (key, value)
                return

        bucket.append((key, value))
        self._size += 1

        if self._size / self._capacity > LOAD_FACTOR_THRESHOLD:
            self._resize()

    def get(self, key):
        idx = self._hash(key)
        bucket = self._buckets[idx]

        for k, v in bucket:
            if k == key:
                return v

        raise KeyError(key)

    def remove(self, key):
        idx = self._hash(key)
        bucket = self._buckets[idx]

        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket.pop(i)
                self._size -= 1
                return v

        raise KeyError(key)

    def contains(self, key):
        try:
            self.get(key)
            return True
        except KeyError:
            return False

    def keys(self):
        all_keys = []
        for bucket in self._buckets:
            for k, v in bucket:
                all_keys.append(k)
        return all_keys

    def values(self):
        all_vals = []
        for bucket in self._buckets:
            for k, v in bucket:
                all_vals.append(v)
        return all_vals

    def size(self):
        return self._size

    def _hash(self, key):
        return hash(key) % self._capacity

    def _resize(self):
        old_buckets = self._buckets
        self._capacity *= 2
        self._buckets = [[] for _ in range(self._capacity)]
        self._size = 0

        for bucket in old_buckets:
            for key, value in bucket:
                self.put(key, value)

    def __repr__(self):
        pairs = []
        for bucket in self._buckets:
            for k, v in bucket:
                pairs.append(f"{k!r}: {v!r}")
        return "HashMap({" + ", ".join(pairs) + "})"

File: javascript/data_structures/HashMap.js
/**
 * Hash Map (separate chaining)
 *
 * Put/Get/Remove: O(1) average, O(n) worst
 *
 * Resizes when load factor > 0.75. Using a simple hash function
 * that works for strings and numbers — not cryptographic obviously.
 */

const INITIAL_CAPACITY = 16;
const LOAD_FACTOR = 0.75;

class HashMap {
  constructor(capacity = INITIAL_CAPACITY) {
    this._capacity = capacity;
    this._size = 0;
    this._buckets = new Array(capacity).fill(null).map(() => []);
  }

  put(key, value) {
    const idx = this._hash(key);
    const bucket = this._buckets[idx];

    // check if key exists and update
    for (let i = 0; i < bucket.length; i++) {
      if (bucket[i][0] === key) {
        bucket[i][1] = value;
        return;
      }
    }

    bucket.push([key, value]);
    this._size++;

    if (this._size / this._capacity > LOAD_FACTOR) {
      this._resize();
    }
  }

  get(key) {
    const idx = this._hash(key);
    const bucket = this._buckets[idx];

    for (const [k, v] of bucket) {
      if (k === key) return v;
    }

    throw new Error(`Key not found: ${key}`);
  }

  remove(key) {
    const idx = this._hash(key);
    const bucket = this._buckets[idx];

    for (let i = 0; i < bucket.length; i++) {
      if (bucket[i][0] === key) {
        const val = bucket[i][1];
        bucket.splice(i, 1);
        this._size--;
        return val;
      }
    }

    throw new Error(`Key not found: ${key}`);
  }

  contains(key) {
    try {
      this.get(key);
      return true;
    } catch {
      return false;
    }
  }

  keys() {
    const result = [];
    for (const bucket of this._buckets) {
      for (const [k] of bucket) {
        result.push(k);
      }
    }
    return result;
  }

  values() {
    const result = [];
    for (const bucket of this._buckets) {
      for (const [, v] of bucket) {
        result.push(v);
      }
    }
    return result;
  }

  size() {
    return this._size;
  }

  _hash(key) {
    const str = String(key);
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      // djb2-ish hash
      hash = (hash * 31 + str.charCodeAt(i)) | 0;
    }
    return Math.abs(hash) % this._capacity;
  }

  _resize() {
    const oldBuckets = this._buckets;
    this._capacity *= 2;
    this._buckets = new Array(this._capacity).fill(null).map(() => []);
    this._size = 0;

    for (const bucket of oldBuckets) {
      for (const [key, value] of bucket) {
        this.put(key, value);
      }
    }
  }
}

module.exports = HashMap;
